filebeat/prospector/log/harvester.go

```diff
@@ -111,3 +125,381 @@ func (h *Harvester) open() error {
+				filesTruncated.Add(1)
+			case ErrRemoved:
+				logp.Info("File was removed: %s. Closing because close_removed is enabled.", h.state.Source)
+			case ErrRenamed:
+				logp.Info("File was renamed: %s. Closing because close_renamed is enabled.", h.state.Source)
+			case ErrClosed:
+				logp.Info("Reader was closed: %s. Closing.", h.state.Source)
+			case io.EOF:
+				logp.Info("End of file reached: %s. Closing because close_eof is enabled.", h.state.Source)
+			case ErrInactive:
+				logp.Info("File is inactive: %s. Closing because close_inactive of %v reached.", h.state.Source, h.config.CloseInactive)
+			default:
+				logp.Err("Read line error: %s; File: ", err, h.state.Source)
+			}
+			return nil
+		}
+
+		// Strip UTF-8 BOM if beginning of file
+		// As all BOMS are converted to UTF-8 it is enough to only remove this one
+		if h.state.Offset == 0 {
+			message.Content = bytes.Trim(message.Content, "\xef\xbb\xbf")
+		}
+
+		// Get copy of state to work on
+		// This is important in case sending is not successful so on shutdown
+		// the old offset is reported
+		state := h.getState()
+		state.Offset += int64(message.Bytes)
+
+		// Create state event
+		data := util.NewData()
+		if h.source.HasState() {
+			data.SetState(state)
+		}
+
+		text := string(message.Content)
+
+		// Check if data should be added to event. Only export non empty events.
+		if !message.IsEmpty() && h.shouldExportLine(text) {
+
+			data.Event = common.MapStr{
+				"@timestamp": common.Time(message.Ts),
+				"source":     state.Source,
+				"offset":     state.Offset, // Offset here is the offset before the starting char.
+			}
+			data.Event.DeepUpdate(message.Fields)
+
+			// Check if json fields exist
+			var jsonFields common.MapStr
+			if fields, ok := data.Event["json"]; ok {
+				jsonFields = fields.(common.MapStr)
+			}
+
+			if h.config.JSON != nil && len(jsonFields) > 0 {
+				reader.MergeJSONFields(data.Event, jsonFields, &text, *h.config.JSON)
+			} else if &text != nil {
+				if data.Event == nil {
+					data.Event = common.MapStr{}
+				}
+				data.Event["message"] = text
+			}
+		}
+
+		// Always send event to update state, also if lines was skipped
+		// Stop harvester in case of an error
+		if !h.sendEvent(data) {
+			return nil
+		}
+		// Update state of harvester as successfully sent
+		h.state = state
+	}
+}
+
+// stop is intended for internal use and closed the done channel to stop execution
+func (h *Harvester) stop() {
+	h.stopOnce.Do(func() {
+		close(h.done)
+	})
+}
+
+// Stop stops harvester and waits for completion
+func (h *Harvester) Stop() {
+	h.stop()
+	h.stopWg.Wait()
+}
+
+// sendEvent sends event to the spooler channel
+// Return false if event was not sent
+func (h *Harvester) sendEvent(data *util.Data) bool {
+	if h.source.HasState() {
+		h.states.Update(data.GetState())
+	}
+
+	err := h.forwarder.Send(data)
+	return err == nil
+}
+
+// sendStateUpdate send an empty event with the current state to update the registry
```

Comment:
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
comment on exported method Harvester.SendStateUpdate should be of the form "SendStateUpdate ..."